[CL-3335] Don't delete User.custom_field_values data when non-user-related custom fields or options deleted

jinjagit opened the pull request:
High priority: to avoid losing (more) production data.

# Bugfix
Fixes 2 related issues:
1. When a `CustomField` with `resource_type: 'Custom_Form'` was deleted, it would cause the deletion of key-value pairs from `User.custom_field_values` where the key values collided.
2. When  `CustomFieldOption` for a `CustomField` with `resource_type: 'Custom_Form'` was deleted, it would cause deletion of values that collided with the `CustomFieldOption.key` if the key also collided with `CustomField.key` (if this was the only value, this would also remove the key from `User.custom_field_values`).

# Notes
I think this is what recently happened with CQC (see ticket [CL-3293](https://citizenlab.atlassian.net/browse/CL-3293)).

I noticed that 2 `CustomFields` with `resource_type: 'CustomForm'` were created and then deleted on the same day on CQC, with _the same key value_ as the `CustomField` with `resource_type: 'User'` that is used by all the Smart Groups that appear to have lost many members (according to the customer).

It would actually be useful if this is the case, since the 2 creation/deletion events were on 27 Mar 2023, and thus the data should still be recoverable from our regular AWS backups.

EDIT: It does, indeed, look like this was the case. I pulled the AWS backup from the day before the customer created then deleted an identically named `CustomField` with `resource_type: 'CustomForm'`, and the backup contained 11k+ more users with the key-value pair in their `custom_field_values` (now restored).

# Changelog
## Fixed
- [CL-3335] Bug where creating then deleting a survey question &/or answer option could cause deletion of user answers to same named registration question &/or answer option.

[CL-3335]: https://citizenlab.atlassian.net/browse/CL-3335?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[CL-3293]: https://citizenlab.atlassian.net/browse/CL-3293?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jinjagit committed:
[CL-3335] Bugfix and regression tests

jinjagit commented on back/app/services/side_fx_custom_field_option_service.rb:
```diff
@@ -16,7 +16,10 @@ def after_update(custom_field_option, current_user)
   def before_destroy(custom_field_option, _current_user)
-    UserCustomFieldService.new.delete_custom_field_option_values custom_field_option.key, custom_field_option.custom_field
+    custom_field = custom_field_option.custom_field
+    return unless custom_field.resource_type == 'User'
```
Extra insurance, just in case we introduce other `CustomField` types, and then include a call to this `before_destroy` method

sebastienhoorens replied:
I think it would be better to move the logic of line 20 inside the helper methods to do the cleanup, as I would consider it part of the cleanup logic. Perhaps we should have a custom field service rather than a user custom field service.

jinjagit committed:
[CL-3335] Skip 2 failing tests (and ask James about them)

jinjagit commented on back/spec/services/timeline_service_spec.rb:
```diff
@@ -66,12 +66,12 @@
     end
 
-    it 'returns the last budgeting phase of the project if there is no currently active phase' do
+    it 'returns the last budgeting phase of the project if there is no currently active phase', skip: 'Failing, why?' do
```
@jamesspeake

These 2 tests fail on this PR, but I can't for the life of me see how any of the changes made here cause this.

I've skipped them for now, since getting this bugfix out seems a priority, but obviously I'd prefer not to leave these tests 'skipped'.

Any ideas?

jamesspeake replied:
@jinjagit Not from looking at the PR, the tests seem to be unrelated to the changes you've made. I'd have to look at it in more detail.

jamesspeake replied:
Possibly something going on in the factories to do with custom fields?

jinjagit replied:
🤔 OK, thanks for looking.

jinjagit replied:
I just followed Sebi's advice from his review, and the tests magically work again! 😄 Now un-skipped.

jinjagit commented on back/engines/commercial/idea_custom_fields/app/controllers/idea_custom_fields/web_api/v1/admin/idea_custom_fields_controller.rb:
```diff
@@ -148,7 +148,6 @@ def update_field!(field, field_params, errors, index)
     end
 
     def delete_field!(field)
-      SideFxCustomFieldService.new.before_destroy field, current_user
```
A main bugfix (of 2).

Since `before_destroy` only calls `UserCustomFieldService.new.delete_custom_field_option_values`, which is only concerned with removing `User.custom_field_values` (both the key & the value(s)), we now don't call it.

sebastienhoorens replied:
I think it's important to keep calling the side fx service, so that we can rely that the right side fx are executed in all situations. If we would want to add some different logic to the delete sidefx, we would probably introduce a bug.

sebastienhoorens reviewed (commented):
Great that you spotted this, thanks! It's annoying, but I wouldn't consider it as urgent, since we had this issue for a while and it's not so likely for keys to collide. 

I feel that dealing with missing keys in general is a complex issue that needs some thinking and good abstraction to properly fix. I added some comments, but if you don't want to spend more time on it, perhaps it could make sense for the method squad to look into this in more detail.

jinjagit commented:
> Great that you spotted this, thanks! It's annoying, but I wouldn't consider it as urgent, since we had this issue for a while and it's not so likely for keys to collide.

@sebastienhoorens 

It might seem that way, but I think it's actually likely (and has happened):

I add a 'what sector are you in?' registration question.
I then create a survey and think 'let's add the same question in the survey'. Then I delete the survey, or remove the question from the survey.

I think this [recently happened on CQC](https://citizenlab.atlassian.net/browse/CL-3293) and resulted in loss of 11k key-values from `User.custom_field_values` records (which I am about to replace after pulling a backup, etc). This broke many of their main smart groups.

jinjagit committed:
[CL-3335] Revert changes to idea_custom_fields_controller.rb

jinjagit committed:
[CL-3335] Move guard clauses into UserCustomFieldService methods

jinjagit committed:
[CL-3335] Remove skipping of 2 tests in timeline_service_spec

jinjagit commented:
@sebastienhoorens Thanks for the review! I have implemented the main changes you suggested:

- I reverted the changes to `idea_custom_fields_contoller` so that the calls to `before_destroy` are replaced.
- I moved the check of `CustomField.resource_type == 'User'` into the helper methods in `UserCustomFieldService`

I also reverted the skipping of 2 tests in the `timeline_service_spec` which magically started working again after the above changes 😄 

I did not, however, rename `UserCustomFieldService` to `CustomFieldService` or do any other more fundamental refactoring.

We can have a quick call to discuss what further work should come out of this bugfix / what tickets to create, but I'd really like to get this out a.s.a.p., if possible, to prevent more data loss on production.

sebastienhoorens reviewed (approved):
All good :) Thanks for solving this!
